### PR TITLE
improve memory footprint of torch.testing.assert_close

### PR DESCRIPTION
**Pull request by pmeier**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #96131

Redo of #90172 out of stack.

### CODE REVIEW

File: torch/testing/_comparison.py [hunk 1]
```diff
@@ -1008,28 +1013,10 @@ def _compare_regular_values_close(
             )
         else:
             msg = make_tensor_mismatch_msg(
-                actual, expected, ~matches, rtol=rtol, atol=atol, identifier=identifier
```
Comment (pmeier) [inline torch/testing/_comparison.py:1011]: In the error case, we created the `mismatches = ~matches` tensor here and turned it back into a `matches` inside `make_tensor_mismatch_msg`. With a minor refactoring, we no longer need to invert `matches` and can use it directly.

File: torch/testing/_comparison.py [hunk 2]
```diff
@@ -991,7 +997,6 @@ def _compare_regular_values_close(
         identifier: Optional[Union[str, Callable[[str], str]]] = None,
     ) -> None:
         """Checks if the values of two tensors are close up to a desired tolerance."""
-        actual, expected = self._promote_for_comparison(actual, expected)
```
Comment (pmeier) [inline torch/testing/_comparison.py:994]: We unconditionally upcasted here in the past, since that was needed for `isclose`. This is no longer the case and so we can just drop that.

File: torch/testing/_comparison.py [hunk 3]
```diff
@@ -260,34 +260,40 @@ def make_tensor_mismatch_msg(
...
+    actual_flat = actual.flatten()
+    expected_flat = expected.flatten()
+    matches_flat = matches.flatten()
 
-    abs_diff = torch.abs(a_flat - b_flat)
+    if not actual.dtype.is_floating_point and not actual.dtype.is_complex:
+        # TODO: Instead of always upcasting to int64, it would be sufficient to cast to the next higher dtype to avoid
+        #  overflow
+        actual_flat = actual_flat.to(torch.int64)
+        expected_flat = expected_flat.to(torch.int64)
```
Comment (pmeier) [inline torch/testing/_comparison.py:289]: However, we still need to upcast in the error case, since we want to display the absolute diff and that is not supported for `torch.bool` and might overflow for other integer dtypes.

File: torch/testing/_comparison.py [hunk 4]
```diff
@@ -260,34 +260,40 @@ def make_tensor_mismatch_msg(
...
     extra = (
         f"Mismatched elements: {total_mismatches} / {number_of_elements} "
         f"({total_mismatches / number_of_elements:.1%})"
     )
 
-    a_flat = actual.flatten()
-    b_flat = expected.flatten()
-    matches_flat = ~mismatches.flatten()
+    actual_flat = actual.flatten()
+    expected_flat = expected.flatten()
```
Comment (pmeier) [inline torch/testing/_comparison.py:282]: Driveby renaming. `a` and `b` were only used in the beginning and should be `actual` and `expected` now.

File: torch/testing/_comparison.py [hunk 5]
```diff
@@ -949,13 +955,13 @@ def _compare_sparse_compressed_values(
             )
 
         self._compare_regular_values_equal(
-            compressed_indices_method(actual),
-            compressed_indices_method(expected),
+            compressed_indices_method(actual).to(torch.int64),
+            compressed_indices_method(expected).to(torch.int64),
```
Comment (pmeier) [inline torch/testing/_comparison.py:959]: Compressed and plain indices in the CSR / CSC / BSR / BSC can be `torch.int32` _or_ `torch.int64`. While the same dtype is enforced within a single tensor, it can be different for two tensors. Thus, we need to convert them to the same dtype here, or `isclose` bail out.

This behavior is the same as before where all integer comparisons where upcasted to int64, but now we need to do it explicitly. Ideally, we wouldn't do that if both inputs are int32, but if it was not an issue before, I prefer to keep the implementation simple here.

Comment (pearu) [reply torch/testing/_comparison.py:959]: Considering the aim of this PR, that is, to reduce the memory footprint, it might be worth finding the maximal dtype of indices, say:
```python
        indices_dtype = compressed_indices_method(actual).dtype
        expected_indices_dtype = compressed_indices_method(expected).dtype
        indices_dtype = indices_dtype if indices_dtype == expected_indices_dtype else torch.int64
        self._compare_regular_values_equal(
            compressed_indices_method(actual).to(indices_dtype),
            compressed_indices_method(expected).to(indices_dtype),
            identifier=f"Sparse {format_name} {compressed_indices_method.__name__}",
        )
```

Comment (pearu) [reply torch/testing/_comparison.py:959]: The found `indices_dtype` can be used for `plain_indices` as well.

File: torch/testing/_comparison.py [hunk 6]
```diff
@@ -260,34 +260,40 @@ def make_tensor_mismatch_msg(
...
+        actual_flat = actual_flat.to(torch.int64)
+        expected_flat = expected_flat.to(torch.int64)
+
+    abs_diff = torch.abs(actual_flat - expected_flat)
     # Ensure that only mismatches are used for the max_abs_diff computation
     abs_diff[matches_flat] = 0
     max_abs_diff, max_abs_diff_flat_idx = torch.max(abs_diff, 0)
 
-    rel_diff = abs_diff / torch.abs(b_flat)
+    rel_diff = abs_diff / torch.abs(expected_flat)
```
Comment (pearu) [inline torch/testing/_comparison.py:296]: A slight OT suggestion: could we have a better normalization factor here (say, `(torch.abs(actual) + torch.abs(expected)) / 2`) for the case where `expected` contains zeros (having zeros is typical, say when comparing the indices of sparse tensors)? Atm, the mismatch messages from `assert_close`  depends on the order of inputs, for example:
```
>>> torch.testing.assert_close(torch.tensor([1, 0]), torch.tensor([1, 1]))
<snip>
Mismatched elements: 1 / 2 (50.0%)
Greatest absolute difference: 1 at index (1,)
Greatest relative difference: 1.0 at index (1,)
>>> torch.testing.assert_close(torch.tensor([1, 1]), torch.tensor([1, 0]))
<snip>
Mismatched elements: 1 / 2 (50.0%)
Greatest absolute difference: 1 at index (1,)
Greatest relative difference: inf at index (1,)
```
(btw, reporting relative differences for non-float tensors is often pointless as well).

**Review by pearu — APPROVED** [review]
LGTM! Thanks, @pmeier!

I have an OT feature request, so feel free to ignore it.

File: torch/testing/_comparison.py (hunk 6 above)
Comment (pmeier) [reply torch/testing/_comparison.py:296]: > Atm, the mismatch messages from `assert_close` depends on the order of inputs, for example:

It's not just the messages, it is the actual op. Internally, we rely on `torch.isclose` and that is already asymmetric. It [defines closeness](https://pytorch.org/docs/stable/generated/torch.isclose.html?highlight=isclose#torch.isclose) as `abs(actual - expected) <= atol + rtol * abs(expected)`. Believe me when I say, we (`torch.testing` team) wanted to change that, but there is just too much inertia. PyTorch is not an outlier here; [`numpy` ](https://numpy.org/doc/stable/reference/generated/numpy.isclose.html#numpy-isclose) (and virtually every other array library) is doing the same.

[Pythons `math` module](https://docs.python.org/3/library/math.html?highlight=isclose#math.isclose) is doing the more sensible thing in defining closeness as `abs(actual - expected) <= max(atol, rtol * max(abs(actual), abs(expected)))`. You can read more about this whole issue in [PEP485](https://peps.python.org/pep-0485/).

At some point we tried to get this behavior specified by the Array API, but couldn't gain enough traction. See data-apis/array-api#170.

> (btw, reporting relative differences for non-float tensors is often pointless as well).

Doesn't that somewhat contradict the use case you gave earlier?

> for the case where expected contains zeros (having zeros is typical, say when comparing the indices of sparse tensors)

**Review by mruberry — APPROVED** [review]

File: test/test_mps.py [hunk 7]
```diff
@@ -10070,7 +10070,7 @@ def test_mps_compat(self):
         # If this test is successful, that means that all operations in the comparison logic are supported natively on
         # the MPS backend. Please remove this test as well as the compatibility logic in
         # torch.testing._comparison.TensorLikePair._equalize_attributes
-        actual = torch.tensor(1.0, device="mps")
+        actual = torch.zeros(2, 3, 4, 5, device="mps")
```
Comment (pmeier) [inline test/test_mps.py:10073]: @kulinseth I've increased the shape to 4 dimensions here, because otherwise this test would pass although `torch.testing.assert_close` is not ready.

Comment (kulinseth) [reply test/test_mps.py:10073]: Sounds good

**Comment by pmeier** [comment]
@kulinseth It seems the test is still passing: https://hud.pytorch.org/pr/96131#12075596190. Does that mean the behavior was fixed? Otherwise, could you send me a patch that consistently makes this test fail? I don't have access to a MPS machine and don't want to waste CI resources by pushing multiple times just for this one test.

**Comment by pmeier** [comment]
@kulinseth any update on this?

**Comment by kulinseth** [comment]
> @kulinseth any update on this?

Sorry for delay @pmeier . I think we have support till 4 dims, if we increase the dimensions to 5 , then test starts failing .

**Comment by pmeier** [comment]
Argh, my bad. Let me fix that.

**Comment by pmeier** [comment]
@kulinseth Test fails now, but unfortunately, the error is not recoverable

```
test_mps.py::TestNoRegression::test_mps_compat Assertion failed: (0 <= mpsAxis && mpsAxis < 4 && "Runtime canonicalization must simplify reduction axes to minor 4 dimensions."), function encodeNDArrayOp, file GPUReductionOps.mm, line 76.
Fatal Python error: Aborted
```

and thus we never hit the xfail. Not sure what to do with this. I'll remove the test to unblock and leave a comment in #95538. LMK if you want to handle it differently.
